### PR TITLE
Chain scripts with promises so they load correctly

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

If we load too fast we actually get an obtuse error that our logic script loaded before our vendor script.

This makes sure they load in the right order

### CODE REVIEW

File: desktop/renderer/renderer.html [hunk 1]
```diff
@@ -2,22 +2,30 @@
...
-      require('electron').ipcRenderer.once('load', (event, options) => {
-        document.title = options.title || document.title
-        options.scripts.forEach(src => {
+      function loadScript (src) {
+        return new Promise((resolve, reject) => {
           const script = document.createElement('script')
-          script.src = src
-          script.onload = () => window.load && window.load(options)
-          document.head.appendChild(script)
+            script.src = src
```
Comment (chrisnojima) [inline desktop/renderer/renderer.html]: just curious if you set async/defer false on the script if it just works

Comment (MarcoPolo) [inline desktop/renderer/renderer.html]: setting async false fixes it too, I'll use that. I won't make the change to plumb the data yet until we need it

Comment (chrisnojima) [inline desktop/renderer/renderer.html]: but we want async false on vendor but async true on our scripts

File: desktop/app/main-window.js [hunk 2]
```diff
@@ -34,8 +34,15 @@ export default function () {
...
       scripts: [
-        ...(__DEV__ ? [resolveRootAsURL('dist', 'dll/dll.vendor.js')] : []),
-        ...[hotPath('index.bundle.js')]],
+        ...(__DEV__ ? [{
+          src: resolveRootAsURL('dist', 'dll/dll.vendor.js'),
+          async: false,
+        }] : []),
+        {
+          src: hotPath('index.bundle.js'),
+          async: !__DEV__,
```
Comment (chrisnojima) [inline desktop/app/main-window.js]: why isn't this async: true?

File: desktop/renderer/renderer.html [hunk 3]
```diff
@@ -2,22 +2,34 @@
...
+          defer != null && (script.defer = defer)
+          script.onload = resolve
+          script.onerror = reject
           document.head.appendChild(script)
         })
+      }
+
+      require('electron').ipcRenderer.once('load', (event, options) => {
+        document.title = options.title || document.title
+        Promise.all(options.scripts.map(opts => loadScript(opts)))
```
Comment (chrisnojima) [inline desktop/renderer/renderer.html]: we don't need this promise at all now right

File: desktop/app/main-window.js (hunk 2 above)
Comment (MarcoPolo) [inline desktop/app/main-window.js]: It'll run into the original error with `async: true`

File: desktop/renderer/renderer.html [hunk 4]
```diff
@@ -12,10 +12,16 @@
...
       require('electron').ipcRenderer.once('load', (event, options) => {
         document.title = options.title || document.title
-        options.scripts.forEach(src => {
+        options.scripts.map(({src, async, defer}) => {
           const script = document.createElement('script')
-          script.src = src
+          if (src == null) {
+            throw new Error('No src for script. Nothing to load')
+          }
+          (script.src = src)
```
Comment (chrisnojima) [inline desktop/renderer/renderer.html]: remove parens

**Comment by MarcoPolo** [comment]
Merging before test finish because last commit was cosmetic

**Pull request merged**